### PR TITLE
Improve stream termination, highWaterMark, objectMode

### LINKED ISSUE #1: `objectMode` should remain `false` when no input streams use `objectMode`
`objectMode: true` is useful when streams handle other types than strings/`Buffer`/`Uint8Array`.
However, `objectMode: false` is more efficient with those types, as it has a higher `highWaterMark` and buffers them better.

When all streams passed to `merge-streams` use `objectMode: false`, the resulting stream should use `objectMode: false` too.

Also, it is somewhat unexpected that merging streams should change their `objectMode`.

### LINKED ISSUE #2: `highWaterMark` should be higher
The `readableHighWaterMark` of the merged stream should be the maximum of the input streams' `readableHighWaterMark`.

For example, when merging one stream producing 8KB chunks and another producing 16KB chunks, the resulting stream should produce 16KB chunks. Otherwise, the merged stream might become a bottleneck, reducing the throughput of the stream pipeline. 

When it comes to the `writableHighWaterMark`, it should be `0`. Since we're using a `PassThrough`, writes complete right away and do not need to be buffered.

### LINKED ISSUE #3: Stream aborts are not propagated
When all the input streams successfully end, so does the merged stream.
When any of the input streams errors, so does the merged stream.

However, when any of the input streams aborts (for example with `stream.destroy()`), the merged stream is not aborted. What's more, the merged stream never ends.

Generally speaking, `finished()` should be used, as it handles [many edge cases](https://github.com/nodejs/node/blob/main/lib/internal/streams/end-of-stream.js) of stream termination, which `merge-streams` does not correctly handle at the moment such as emitting the `end` event twice.

### PR DESCRIPTION
**Pull request by ehmicky**
Fixes #1 
Fixes #2 
Fixes #3 

Those issues are unrelated. I can break this down into 3 PRs instead if you'd prefer, please let me know :+1: 

Note: those are breaking changes, i.e. would require a major release.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,7 +19,6 @@
 	"engines": {
 		"node": ">=18"
 	},
-	"sideEffects": false,
```
Comment (ehmicky) [inline package.json:22]: This was done automatically by `npm install`. :thinking:

**Comment by sindresorhus** [comment]
Thanks for improving these things. You definitely have a better grasp on Node.js streams than me.

**Comment by sindresorhus** [comment]
I assume only #1 is a breaking change?

File: package.json (hunk 1 above)
Comment (sindresorhus) [reply package.json:22]: Weird. It should not remove custom properties.

Comment (ehmicky) [reply package.json:22]: Oh this is because it is already present in line `18`. That's a duplicate line.

**Comment by ehmicky** [comment]
Potentially changing the `objectMode` (#1) is a breaking change for users who pass streams not in object mode, since now the merged stream will not be in object mode. In practice, they will probably not notice it though (unless they do more advanced logic with the stream).

Potentially changing the `highWaterMark` (#2) could technically result in some consumer's logic behaving differently, if that logic depends on reading the streams a specific amount of times, or if it assumes the chunks being read to have a specific size. I.e. it is technically a breaking change, although unlikely to actually change consumer's code.

Aborting the merged stream when any input stream is aborted (e.g. through `.destroy()`) (#3) is a bug fix.

File: index.js [hunk 2]
```diff
@@ -1,39 +1,58 @@
...
+	stream.pipe(passThroughStream, {end: false});
 
-		stream.on('error', error => {
-			passThroughStream.emit('error', error);
-		});
+	try {
+		await streamPromises;
+		passThroughStream.end();
+	} catch (error) {
+		if (error?.code === 'ERR_STREAM_PREMATURE_CLOSE') {
```
Comment (ehmicky) [inline index.js:52]: That's the error thrown by `finished()` on aborts (e.g. due to `.destroy()`).

https://github.com/nodejs/node/blob/e57a32a03a9682857ed83c38447291b3e83b983e/lib/internal/streams/end-of-stream.js#L154

File: test.js [hunk 3]
```diff
@@ -1,53 +1,136 @@
...
+	while (stream.read() !== null) {
+		readCount += 1;
+		// eslint-disable-next-line no-await-in-loop
+		await scheduler.yield();
+	}
+
+	t.is(readCount, bufferCount);
+
+	t.pass();
+};
```
Comment (ehmicky) [inline test.js:133]: Each `PassThrough` has one read buffer + one write buffer of size `highWaterMark`. The merged stream too, except it does not have a write buffer (`writableHighWaterMark` is `0`).

Therefore, the streams will be full after writing `5 * highWaterMark` bytes.
Then, the merged stream will consume all those bytes in exactly 5 reads, returning each `highWaterMark` bytes.

File: index.js [hunk 4]
```diff
@@ -1,39 +1,56 @@
...
+		.map(({readableHighWaterMark}) => readableHighWaterMark);
+	return Math.max(...highWaterMarks);
+};
+
+const pipeStream = async (passThroughStream, streams) => {
+	try {
+		await Promise.all(streams.map(stream => finished(stream, {cleanup: true, readable: true, writable: false})));
+		passThroughStream.end();
+	} catch (error) {
+		// This is the error thrown by `finished()` on `stream.destroy()`
```
Comment (ehmicky) [inline index.js:49]: https://github.com/nodejs/node/blob/e57a32a03a9682857ed83c38447291b3e83b983e/lib/internal/streams/end-of-stream.js#L154

**Comment by sindresorhus** [comment]
Thanks 🙏

**Pull request merged**